dvc/objects/tree.py

```diff
@@ -115,24 +119,45 @@ def load(cls, odb, hash_info):
         tree = cls.from_list(raw)
         tree.path_info = obj.path_info
         tree.fs = obj.fs
+        for _, entry_obj in tree:
+            entry_obj.fs = obj.fs
         tree.hash_info = hash_info
 
         return tree
 
-    def filter(self, odb, prefix):
+    def filter(
+        self, odb: "ObjectDB", prefix: Tuple[str], copy: bool = False
+    ) -> Optional[HashFile]:
+        """Return filter object(s) for this tree.
+
+        If copy is True, returned object will be a Tree containing
+        filtered entries, but with hash_info copied from the original tree.
+
+        If copy is False, returned object will be a raw HashFile or Tree with
+        newly computed hash_info for the filtered object.
+        """
```

Comment:
The current behavior for `filter` is to return a new tree containing the filtered set of trie entries, and a newly computed digest for the entire tree. These filtered trees are used when we do granular stage/output operations (and are related to https://github.com/iterative/dvc/issues/5988)

For collecting used dir cache, what we really want is a filtered version of `self` which contains the original `self.hash_info`, but with the filtered trie entries.

The use of `copy` is kind of ugly right now, but I'm also not sure that we really need 2 different filter methods?